crates/ra_analysis/src/descriptors/mod.rs

```diff
@@ -61,3 +64,49 @@ impl FnDescriptor {
         res
     }
 }
+
+#[derive(Debug)]
+pub struct ReferenceDescriptor {
+    pub range: TextRange,
+    pub name: String
+}
+
+#[derive(Debug)]
+pub struct DeclarationDescriptor<'a> {
```

Comment:
Descriptors should be owned values and, ideally, should not directly reference syntax trees. 

The idea is that, when processing code, we convert from syntax trees to a descriptor tree (which is more compact) and then throw away syntax to save memory. 

We should have a method to map between descriptors and syntax on as needed basis. For example, we can store some sort of syntax node id in the descriptor, and use it to retrieve the node.

Reply:
I wrestled with this but found that it was really the only way to implement `find_all_refs` without passing in the `syntax`. If we did that then we need to make sure that the syntax is the right one for the stored `TextRange` (maybe by storing the `FileId` as well?).

Can you sketch out a little what this would look like? I'm assuming that we do not have the concept of a syntax node id. If we had one where would it look for the corresponding syntax?

Reply:
>Can you sketch out a little what this would look like? 

I won't have time until next week to give significant advice unfortunately, but I want to document/refactor/flesh-out the descriptor infrastructure more in the coming weeks.

The motivation for descriptors is that there isn't a 1-to-1 correspondance between rust source code and rust semantic model. If you have

```
#[path="a.rs"]
mod foo;
#[path="a.rs"]
mod bar;
```

Then the source of `a.rs` is present in the semantic model as two unrelated modules. 

So, while SyntaxTrees correspond to source files, descriptors should correspond to the semantic model, and most of the analysis should work solely with descriptors.

Practically speaking, I think we should give ids to all syntax nodes and descriptors, and then have a free-form mapping between these ids.

I think a good star might be just using a triple of `(FIleId, TextRange, SyntaxKind)` as a syntax node ID.

Reply:
Makes sense. I'll have to think about how this should work when converting from id to node inside `find_all_refs`. It looks like we'll need to access (pass in?) the `db` somehow.

Reply:
See https://github.com/rust-analyzer/rust-analyzer/pull/169 for the syntax ptr stuff. I'll try to move some of the desciptors to this.